### ISSUE TITLE
FactoryClass changes during the parsing of metadata

### ISSUE THREAD

**Issue by jeremyf** [samvera-labs/bulkrax#702]
During the [Bulkrax::OaiEntry#build\_metadata](https://github.com/samvera-labs/bulkrax/blob/b857a25becefc4bd960f25647033dfb2e899386e/app/models/bulkrax/oai_entry.rb#L28-L47) we check [Bulkrax::HasMatchers#field\_supported?](https://github.com/samvera-labs/bulkrax/blob/b857a25becefc4bd960f25647033dfb2e899386e/app/models/concerns/bulkrax/has_matchers.rb#L123-L129) for each field.  Note the call to the `#factory_class` method.

```ruby
def field_supported?(field)
  field = field.gsub('_attributes', '')
  
  return false if excluded?(field)
  return true if supported_bulkrax_fields.include?(field)
  return factory_class.method_defined?(field) && factory_class.properties[field].present?
end
```

For works, the [Bulkrax::ImportBehavior#factory\_class](https://github.com/samvera-labs/bulkrax/blob/b857a25becefc4bd960f25647033dfb2e899386e/app/models/concerns/bulkrax/import_behavior.rb#L180-L197) definition is contingent on the `parsed_metadata`. 

```ruby
def factory_class
  fc = if self.parsed_metadata&.[]('model').present?
    	 self.parsed_metadata&.[]('model').is_a?(Array) ? self.parsed_metadata&.[]('model')&.first : self.parsed_metadata&.[]('model')
       elsif self.mapping&.[]('work_type').present?
    	 self.parsed_metadata&.[]('work_type').is_a?(Array) ? self.parsed_metadata&.[]('work_type')&.first : self.parsed_metadata&.[]('work_type')
       else
    	 Bulkrax.default_work_type
       end
  
  # return the name of the collection or work
  fc.tr!(' ', '_')
  fc.downcase! if fc.match?(/[-_]/)
  fc.camelcase.constantize
rescue NameError
  nil
rescue
  Bulkrax.default_work_type.constantize
end
```

With the `factory_class` derived from the `parsed_metadata` this means that during an entry’s ingest the `factory_class` can and will almost certainly change.  Thus up until we’ve set the `model` in the `parsed_metadata` we are using the configured default work type.

The impact is that the `field_supported?` method will return erratic results.  Case in point I have found that for Adventist’s import the fields `volume_number` and `pagination` are marked as not supported fields, because they are not on the default work type **and** when we encounter that node we haven’t yet set the `model` for the `parsed_metadata`.

In other words, we are likely to skip over the unique fields fields for the custom works of objects.

This is a bug that almost certainly plagues all XML based parsers.

The proposed solution is to amend the `build_metadata` method to first establish the `factory_class`.

Relates to:

-   <https://github.com/scientist-softserv/adventist-dl/issues/188>
-   <https://github.com/scientist-softserv/adventist-dl/issues/187>
